### ISSUE TITLE
Missing engine. TypeError: Cannot read property 'getUserAgent' of undefined

### ISSUE THREAD

**Issue by nbanguiano** [macbre/phantomas#722]
Just following the initial instructions: install -> run the example
`phantomas https://github.com/macbre/phantomas --verbose`
I get the following error:
`/usr/local/lib/node_modules/phantomas/lib/engines.js:79`
`this.options['user-agent'] = "phantomas/" + VERSION + " (" + engine.getUserAgent() + "; " + process.platform + " " + process.arch + ")";`
`TypeError: Cannot read property 'getUserAgent' of undefined`

So `engine` is undefined, and specifying the `--engine` doesn't helpr.
(I don't have PhantomJs installed and configured.
The README makes it clear that I didn't need to... or do I?)

The debug log, is in the gist below. Specifically
`
phantomas:engines Engine "webkit": failed to import - Error: Cannot find module 'phantomjs-prebuilt-macbre' +0ms`
`phantomas:engines Engine "gecko": failed to import - Error: Cannot find module 'slimerjs' +1ms`
https://gist.github.com/nbanguiano/3d5327e6f139ebe1c5eee29a222c92fa

**Comment by SchnockNicolas** [comment]
same issue on mac. I can make phantomjs work either.
I use a docker instead.

**Comment by macbre** [comment]
@nbanguiano - sorry for the late reply.

Did you run this command to install phantomas?

```
npm install --global --no-optional phantomas phantomjs-prebuilt-macbre
```

It seems that you do not have PhantomJS installed.

**Comment by Mathieu-C** [comment]
@macbre I have the same issue on Mac even though I've run the install command you provided. Trying to install it using Yarn gave me this:

```
➜  ~ yarn global add phantomas phantomjs-prebuilt-macbre
yarn global v1.10.1
warning package.json: No license field
[1/4] 🔍  Resolving packages...
[2/4] 🚚  Fetching packages...
[3/4] 🔗  Linking dependencies...
[4/4] 📃  Building fresh packages...
error /Users/mathieuchatillon/.config/yarn/global/node_modules/phantomjs-prebuilt-macbre: Command failed.
Exit code: 1
Command: node install.js
Arguments:
Directory: /Users/mathieuchatillon/.config/yarn/global/node_modules/phantomjs-prebuilt-macbre
Output:
Considering PhantomJS found at /usr/local/bin/phantomjs
Found PhantomJS at /usr/local/bin/phantomjs ...verifying
PhantomJS detected, but wrong version 2.1.1 @ /usr/local/bin/phantomjs.
Unexpected platform or architecture: darwin/x64
It seems there is no binary available for your platform/architecture
Try to install PhantomJS globally
```
https://gist.github.com/Mathieu-C/75092969949de98fded62f3e42d8412f

**Comment by xiongchiamiov** [comment]
Also seeing this.  If I do an `npm list --global`, it doesn't appear phantomjs was pulled in as a dependency.  There's also a very suspicious message about not having a package.json for phantomjs-prebuilt-macbre.

```
├─┬ phantomas@1.20.1
│ ├─┬ analyze-css@0.12.6
│ │ ├─┬ cli@1.0.1
│ │ │ ├── exit@0.1.2
│ │ │ └── glob@7.1.3 deduped
│ │ ├─┬ css@2.2.4
│ │ │ ├── inherits@2.0.3 deduped
│ │ │ ├── source-map@0.6.1
│ │ │ ├─┬ source-map-resolve@0.5.2
│ │ │ │ ├── atob@2.1.2
│ │ │ │ ├── decode-uri-component@0.2.0
│ │ │ │ ├── resolve-url@0.2.1
│ │ │ │ ├── source-map-url@0.4.0
│ │ │ │ └── urix@0.1.0 deduped
│ │ │ └── urix@0.1.0
│ │ ├── css-shorthand-properties@1.1.1
│ │ ├── debug@3.2.6 deduped
│ │ ├── fast-stats@0.0.3
│ │ ├─┬ glob@7.1.3
│ │ │ ├── fs.realpath@1.0.0
│ │ │ ├─┬ inflight@1.0.6
│ │ │ │ ├── once@1.4.0 deduped
│ │ │ │ └── wrappy@1.0.2
│ │ │ ├── inherits@2.0.3 deduped
│ │ │ ├─┬ minimatch@3.0.4
│ │ │ │ └─┬ brace-expansion@1.1.11
│ │ │ │   ├── balanced-match@1.0.0
│ │ │ │   └── concat-map@0.0.1
│ │ │ ├─┬ once@1.4.0
│ │ │ │ └── wrappy@1.0.2 deduped
│ │ │ └── path-is-absolute@1.0.1
│ │ ├─┬ http-proxy-agent@2.1.0
│ │ │ ├─┬ agent-base@4.2.1
│ │ │ │ └─┬ es6-promisify@5.0.0
│ │ │ │   └── es6-promise@4.2.5
│ │ │ └─┬ debug@3.1.0
│ │ │   └── ms@2.0.0
│ │ ├── node-fetch@2.2.0
│ │ ├── onecolor@3.1.0
│ │ ├── optimist@0.6.1 deduped
│ │ ├── slick@1.12.2
│ │ └── specificity@0.3.2
│ ├── ansicolors@0.3.2
│ ├── ansistyles@0.1.3
│ ├── ascii-table@0.0.9
│ ├─┬ async@2.6.1
│ │ └── lodash@4.17.11
│ ├── csv-string@3.1.5
│ ├─┬ debug@3.2.6
│ │ └── ms@2.1.1
│ ├─┬ js-yaml@3.12.0
│ │ ├─┬ argparse@1.0.10
│ │ │ └── sprintf-js@1.0.3
│ │ └── esprima@4.0.1
│ ├── node-statsd@0.1.1
│ ├─┬ optimist@0.6.1
│ │ ├── minimist@0.0.10
│ │ └── wordwrap@0.0.3
│ ├─┬ optimist-config-file@1.0.1
│ │ ├── ansistyles@0.1.3 deduped
│ │ ├─┬ debug@2.6.9
│ │ │ └── ms@2.0.0
│ │ ├── js-yaml@3.12.0 deduped
│ │ └── optimist@0.6.1 deduped
│ ├── UNMET OPTIONAL DEPENDENCY phantomjs-prebuilt-macbre@^2.5.0-dev2
│ ├── progress@2.0.1
│ ├── q@1.5.1
│ ├── UNMET OPTIONAL DEPENDENCY slimerjs@^0.10.3
│ ├─┬ tap-producer-macbre@0.0.3
│ │ ├── inherits@2.0.3
│ │ └── yamlish@0.0.7
│ ├── travis-fold@0.1.2
│ └── uuid@3.3.2
└─┬  error: ENOENT: no such file or directory, open '/usr/local/lib/node_modules/phantomjs-prebuilt-macbre/package.json
  ├── assert-plus@1.0.0
  ├─┬ combined-stream@1.0.7
  │ └── delayed-stream@1.0.0
  ├── core-util-is@1.0.2
  ├── extsprintf@1.3.0
  ├── inherits@2.0.3
  ├── jsbn@0.1.1
  ├─┬ mime-types@2.1.21
  │ └── mime-db@1.37.0
  ├── safe-buffer@5.1.2
  ├── safer-buffer@2.1.2
  └── tweetnacl@0.14.5
```

When I initially did the install, I didn't get any errors:

```
[$]> npm install --global --no-optional phantomas phantomjs-prebuilt-macbre
/usr/local/bin/phantomas -> /usr/local/lib/node_modules/phantomas/bin/phantomas.js
+ phantomas@1.20.1
+ phantomjs-prebuilt-macbre@2.5.0-dev2
added 72 packages from 138 contributors in 6.111s
```

but when I run the install again only for phantomjs, I get an error saying that there's no prebuilt binary for mac:

```
[$]> npm install --global --no-optional phantomjs-prebuilt-macbre
/usr/local/bin/phantomjs -> /usr/local/lib/node_modules/phantomjs-prebuilt-macbre/bin/phantomjs

> phantomjs-prebuilt-macbre@2.5.0-dev2 install /usr/local/lib/node_modules/phantomjs-prebuilt-macbre
> node install.js

Considering PhantomJS found at /usr/local/bin/phantomjs
Looks like an `npm install -g`
Could not link global install, skipping...
Unexpected platform or architecture: darwin/x64
It seems there is no binary available for your platform/architecture
Try to install PhantomJS globally
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! phantomjs-prebuilt-macbre@2.5.0-dev2 install: `node install.js`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the phantomjs-prebuilt-macbre@2.5.0-dev2 install script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/jamesp/.npm/_logs/2018-10-31T21_40_13_865Z-debug.log
```

Here's the last part of that debug log:
```
1969 info lifecycle phantomjs-prebuilt-macbre@2.5.0-dev2~install: phantomjs-prebuilt-macbre@2.5.0-dev2
1970 verbose lifecycle phantomjs-prebuilt-macbre@2.5.0-dev2~install: unsafe-perm in lifecycle true
1971 verbose lifecycle phantomjs-prebuilt-macbre@2.5.0-dev2~install: PATH: /usr/local/lib/node_modules/npm/node_modules/npm-lifecycle/node-gyp-bin:/usr/local/lib/node_modules/phantomjs-prebuilt-macbre/node_modules/.bin:/usr/local/lib/node_modules/.bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/usr/local/sbin:/usr/local/MacGPG2/bin:/Users/jamesp/bin:/Users/jamesp/we-tools-cli/bin
1972 verbose lifecycle phantomjs-prebuilt-macbre@2.5.0-dev2~install: CWD: /usr/local/lib/node_modules/phantomjs-prebuilt-macbre
1973 silly lifecycle phantomjs-prebuilt-macbre@2.5.0-dev2~install: Args: [ '-c', 'node install.js' ]
1974 silly lifecycle phantomjs-prebuilt-macbre@2.5.0-dev2~install: Returned: code: 1  signal: null
1975 info lifecycle phantomjs-prebuilt-macbre@2.5.0-dev2~install: Failed to exec install script
1976 timing action:install Completed in 282ms
1977 verbose unlock done using /Users/jamesp/.npm/_locks/staging-3a08f0df5026584d.lock for /usr/local/lib/node_modules/.staging
1978 timing stage:rollbackFailedOptional Completed in 90ms
1979 timing stage:runTopLevelLifecycles Completed in 1979ms
1980 verbose stack Error: phantomjs-prebuilt-macbre@2.5.0-dev2 install: `node install.js`
1980 verbose stack Exit status 1
1980 verbose stack     at EventEmitter.<anonymous> (/usr/local/lib/node_modules/npm/node_modules/npm-lifecycle/index.js:301:16)
1980 verbose stack     at EventEmitter.emit (events.js:182:13)
1980 verbose stack     at ChildProcess.<anonymous> (/usr/local/lib/node_modules/npm/node_modules/npm-lifecycle/lib/spawn.js:55:14)
1980 verbose stack     at ChildProcess.emit (events.js:182:13)
1980 verbose stack     at maybeClose (internal/child_process.js:962:16)
1980 verbose stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:251:5)
1981 verbose pkgid phantomjs-prebuilt-macbre@2.5.0-dev2
1982 verbose cwd /Users/jamesp/Documents
1983 verbose Darwin 16.7.0
1984 verbose argv "/usr/local/Cellar/node/11.0.0/bin/node" "/usr/local/bin/npm" "install" "--global" "--no-optional" "phantomjs-prebuilt-macbre"
1985 verbose node v11.0.0
1986 verbose npm  v6.4.1
1987 error code ELIFECYCLE
1988 error errno 1
1989 error phantomjs-prebuilt-macbre@2.5.0-dev2 install: `node install.js`
1989 error Exit status 1
1990 error Failed at the phantomjs-prebuilt-macbre@2.5.0-dev2 install script.
1990 error This is probably not a problem with npm. There is likely additional logging output above.
1991 verbose exit [ 1, true ]
```

**Comment by shervud** [comment]
Hi, here is a working solution for me

Open phantomas/node_modules/lib/engines/webkit.js and change 
require('phantomjs-prebuilt-macbre') to require('phantomjs-prebuilt') and of course install **"phantomjs-prebuilt"**

However it is strange, why  'phantomjs-prebuilt-macbre' does not work having in mind it has been installed, may be due to some symlinks and so on...

**Comment by kkmoslehpour** [comment]
I'm seeing this issue as well in Centos 7, was there a solution to this finally?

**Comment by macbre** [comment]
This will be solved by phantomas v2.0.0 that always install Puppeteer as a dependency.
